### PR TITLE
[stdlib] [Permutation] constructive pigeonhole principle for lists

### PR DESCRIPTION
**Pull request by mrhaandi**
Kind: feature

This PR adds the constructive pigeonhole principle for lists:
```coq
Lemma Permutation_pigeonhole_rel (R : B -> A -> Prop) (l1 : list B) (l2 : list A) :
  Forall (fun (b : B) => Exists (R b) l2) l1 ->
  length l2 < length l1 ->
  exists (b b' : B) (l3 : list B), Permutation l1 (b :: b' :: l3) /\ exists a, In a l2 /\ R b a /\ R b' a.
```

My motivation is the constructive computability theory, where this lemma is quite useful.
Note: the proof does not require excluded middle, decidable equality, or decidable `R` on the underlying types.
Other formulations in the stdlib are
- `Pigeonhole_principle` from `Sets.Image`, which requires classical logic
- `NoDup_Permutation_bis` from `Sorting.Permutation`, which is dual. It is not sufficient because contrapositive `not (Nodup l)` is constructively too weak.

Lemmas added to `Permutation.v`
- `Permutation_incl_cons_inv_r`
- `Permutation_pigeonhole`
- `Permutation_pigeonhole_rel`

Lemmas added to `List.v`
- `Forall2_cons_iff`
- `Forall2_length`

- [x] Added **changelog**.

### CODE REVIEW

**Comment by DmxLarchey** [comment]
Hi @mrhaandi, 

For the moment, I just add [this GitHub reference](https://github.com/DmxLarchey/PHP-short) in case of further discussions. The length of the code is quite dependent on how you express that a list has a duplicate: with permutations, or of the form "...++x::...++x::..." etc.

Maybe the code could be completed with proofs that double inclusion (and inclusion) of lists can be characterized inductivelly (smallest pre-order allowing contraction, weakening ...) w/o assuming decidable equality.

**Comment by mrhaandi** [comment]
> of the form "...++x::...++x::..." etc.

The existing
```coq
Lemma Permutation_vs_cons_cons_inv : forall l l' a b,
 Permutation l (a :: b :: l') ->
 exists l1 l2 l3, l = l1 ++ a :: l2 ++ b :: l3 \/ l = l1 ++ b :: l2 ++ a :: l3.
```
can be used to easily transition to `...++x::...++x::...`

File: theories/Sorting/Permutation.v
Comment (olaure01) [inline theories/Sorting/Permutation.v:566]:
```suggestion
        now apply PeanoNat.Nat.succ_lt_mono. }
```

File: theories/Sorting/Permutation.v
Comment (olaure01) [inline theories/Sorting/Permutation.v:570]:
```suggestion
      now apply (Permutation_cons_app (b :: b :: nil)).
```

File: theories/Sorting/Permutation.v
Comment (olaure01) [inline theories/Sorting/Permutation.v:568]:
```suggestion
      exists b, (a :: l3).
```

File: theories/Sorting/Permutation.v
Comment (olaure01) [inline theories/Sorting/Permutation.v:577]:
```suggestion
Lemma Permutation_pigeonhole_rel (R : B -> A -> Prop) l1 l2 :
  Forall (fun b => Exists (R b) l2) l1 ->
  length l2 < length l1 ->
  exists b b' l3, Permutation l1 (b :: b' :: l3) /\ exists a, In a l2 /\ R b a /\ R b' a.
```

File: theories/Sorting/Permutation.v
Comment (olaure01) [inline theories/Sorting/Permutation.v:588]:
```suggestion
  intros Hl2l2'. rewrite <- (Forall2_length Hl2'l1) in Hl2l2'.
```

File: theories/Sorting/Permutation.v
Comment (olaure01) [inline theories/Sorting/Permutation.v:586]: Maybe turn this into a lemma:
```coq
Lemma Forall_Exists_exists_Forall2 (R : B -> A -> Prop) l1 l2 :
  Forall (fun b => Exists (R b) l2) l1 ->
  exists l2', Forall2 R l1 l2' /\ incl l2' l2.
```

File: theories/Lists/List.v
Comment (mrhaandi) [reply theories/Lists/List.v]: Done.

File: theories/Lists/List.v
Comment (mrhaandi) [reply theories/Lists/List.v]: Done.

File: theories/Sorting/Permutation.v
Comment (mrhaandi) [reply theories/Sorting/Permutation.v]: Done.

File: theories/Sorting/Permutation.v
Comment (mrhaandi) [reply theories/Sorting/Permutation.v]: Done.

File: theories/Sorting/Permutation.v
Comment (mrhaandi) [reply theories/Sorting/Permutation.v]: Done.

File: theories/Sorting/Permutation.v
Comment (mrhaandi) [reply theories/Sorting/Permutation.v]: Done.

File: theories/Sorting/Permutation.v
Comment (mrhaandi) [reply theories/Sorting/Permutation.v]: Done.

File: theories/Sorting/Permutation.v
Comment (mrhaandi) [reply theories/Sorting/Permutation.v]: Done.

File: theories/Sorting/Permutation.v
Comment (mrhaandi) [reply theories/Sorting/Permutation.v]: Done.

File: theories/Sorting/Permutation.v
Comment (mrhaandi) [reply theories/Sorting/Permutation.v]: This cannot be changed completely, `(l1 : list B)` is necessary because of above `Implicit Types l m : list A.`

**Comment by olaure01** [comment]
@mrhaandi CI is OK, you can squash commits and maybe add `Forall_Exists_exists_Forall2` to changelog.
I will merge by Thursday if there is no comments.

**Review by olaure01 — APPROVED** [review]

**Comment by mrhaandi** [comment]
> CI is OK, you can squash commits and maybe add `Forall_Exists_exists_Forall2` to changelog. I will merge by Thursday if there is no comments.

Done; also I did a fresh rebase to be on the safe side. Thank you, Olivier, for the help.

**Comment by mrhaandi** [comment]
> I would consider easier to merge #16037 first, in case of troubles.

I agree, lets wait for #16037 (planned later today) and I'll rebase afterwards.
In my proofs I did follow strict focusing.
(Also, I really appreciate #16037.)

**Comment by mrhaandi** [comment]
I now rebased on top of master after https://github.com/coq/coq/pull/16037 was merged.

**Comment by Alizter** [comment]
@coq/windows-build-maintainers what is going on in the CI?

**Comment by olaure01** [comment]
@coqbot merge now

**Pull request merged**
